File: services/query/flow/src/Dataset/Catalog.php
<?php

declare(strict_types=1);

namespace Aiwatcher\Flow\Dataset;

use Flow\ETL\DataFrame;
use Nyholm\Psr7\Request;
use Psr\Http\Client\ClientInterface;

use function Flow\ETL\Adapter\Http\from_http_paginated;
use function Flow\ETL\Adapter\Http\http_pagination_cursor;
use function Flow\ETL\Adapter\Http\http_pagination_offset;
use function Flow\ETL\Adapter\Http\http_request_option_query;
use function Flow\ETL\Adapter\Http\http_stop_when_empty_path;
use function Flow\ETL\Adapter\Http\http_stop_when_max_results;
use function Flow\ETL\DSL\array_expand;
use function Flow\ETL\DSL\array_get;
use function Flow\ETL\DSL\cast;
use function Flow\ETL\DSL\data_frame;
use function Flow\ETL\DSL\optional;
use function Flow\ETL\DSL\ref;
use function Flow\Types\DSL\type_array;

/**
 * Every dataset a query can read, and how to turn one into a `DataFrame`.
 *
 * ## Why HTTP and not a local copy
 *
 * Measured on 1500 runs: `groupBy(agent)` over `/api/v1/runs` through Flow's
 * paginated HTTP extractor takes 210 ms. The same question over 175 000 raw
 * events in Parquet takes 2 s, and over the raw write-ahead log 16 s. What
 * decides it is the grain, not the transport — the API already folds events
 * into run summaries, so there are 1500 rows to read instead of 175 000.
 *
 * So there is no export, no ingest job and no second copy of the data. The
 * cursors the API already serves (`next_cursor` in the body, replayed as a
 * query parameter) are exactly the shape `http_pagination_cursor` expects.
 *
 * The cost is that a query only sees what the API serves, which is the read
 * model's retention window. History older than that needs the columnar path;
 * see ADR_0008 for the condition under which that gets built.
 *
 * ## What a dataset is, and where that is written
 *
 * Not here. `services/query/contract/catalog.json` declares every dataset — its
 * route, its columns, its hints, its `read()` arguments — and every query engine
 * loads that one file (AW-3), so the DataFusion and DuckDB engines serve exactly
 * the catalog this one does. What is Flow's is how a declared dataset becomes a
 * `DataFrame`, which is `open()` below.
 */
final readonly class Catalog
{
    /** @var array<string, Dataset> */
    private array $datasets;

    public function __construct(
        private ClientInterface $client,
        private string $baseUrl,
        /** How many rows to ask for per request. Fewer, larger pages beat many small ones. */
        private int $pageSize = 500,
        /**
         * A directory of part files, for the one dataset not served by the API.
         *
         * Null — the default — and the catalog is the API and nothing else. See
         * `Corpus` for why the exception exists and what it may read.
         */
        private ?string $corpusRoot = null,
        /** The declared catalog. Null is the one every engine ships with. */
        ?string $catalogFile = null,
    ) {
        // A corpus dataset is offered only where a corpus root is configured:
        // without one it names files that are not there, and "no spans" would be
        // an answer about the data when the truth is that nobody pointed at any.
        $this->datasets = \array_filter(
            CatalogFile::read($catalogFile ?? CatalogFile::defaultPath()),
            static fn(Dataset $dataset): bool => $dataset->corpus === null || $corpusRoot !== null,
        );
    }

    /**
     * The query string one page of a dataset is read with.
     *
     * The window is the panel's time control, forwarded to the API rather than
     * applied to the rows here: filtering after the fact would page through
     * everything to throw most of it away, and the API already answers the
     * question. Datasets that do not take one are read whole — see
     * `Dataset::$windowed`.
     */
    private static function query(
        Dataset $dataset,
        int $pageSize,
        ?int $windowSeconds,
        array $arguments = [],
        ?int $asOf = null,
    ): string {
        $parameters = ['limit' => $pageSize];

        if ($dataset->windowed && $windowSeconds !== null && $windowSeconds > 0) {
            $parameters['window_seconds'] = $windowSeconds;
        }

        // The end of the window, when a managed plan pinned one. Absent for
        // every panel query, which is what keeps a link somebody pastes meaning
        // "the last hour" at the moment it is opened. Present, the window is a
        // closed span and a retry five minutes later reads the same rows.
        if ($dataset->windowed && $asOf !== null) {
            $parameters['as_of'] = $asOf;
        }

        // Only what the dataset declared. The API rejects unknown query
        // parameters rather than ignoring them, so forwarding whatever a query
        // happened to name would turn a typo into a 400 about the whole read.
        foreach ($dataset->parameters as $name => $_) {
            if (!isset($arguments[$name]) || $arguments[$name] === '') {
                continue;
            }

            $parameters[$name] = $arguments[$name];
        }

        return \http_build_query($parameters);
    }

    /** `default` is `runs`: the grain the explorer shows, and the cheap one. */
    public function resolve(string $name): ?Dataset
    {
        return $this->datasets[$name === 'default' ? 'runs' : $name] ?? null;
    }

    /** @return array<string, Dataset> */
    public function all(): array
    {
        return $this->datasets;
    }

    /**
     * Open a dataset as a `DataFrame` with its columns already flat.
     *
     * The HTTP extractor yields one row per *response*, carrying the raw body
     * and the request that produced it. Three steps turn that into rows: decode
     * the body, explode the array of records into one row each, then project
     * the declared columns. Whoever writes the query sees run columns, never the
     * HTTP envelope.
     */
    /**
     * @param array<string, string> $arguments the read()'s declared named arguments
     * @param ?int                  $asOf      the instant a managed plan pinned its window to end at
     */
    public function open(
        Dataset $dataset,
        ?string $run = null,
        ?int $windowSeconds = null,
        array $arguments = [],
        ?int $asOf = null,
        ?int $inputLimit = null,
    ): DataFrame {
        if ($dataset->corpus !== null && $this->corpusRoot !== null) {
            return (new Corpus($this->corpusRoot))->open($dataset, $arguments, $inputLimit);
        }

        $path = $dataset->requiresRun
            ? \str_replace('{run}', \rawurlencode((string) $run), $dataset->path)
            : $dataset->path;

        $request = new Request(
            'GET',
            $this->baseUrl . $path . '?' . self::query($dataset, $this->pageSize, $windowSeconds, $arguments, $asOf),
        );

        $pagination = http_pagination_cursor('next_cursor', http_request_option_query($dataset->cursorParam));
        $rowLimit = null;
        if ($dataset->name === 'hub_rows') {
            $rowLimit = \filter_var($arguments['limit'] ?? '100', \FILTER_VALIDATE_INT, ['options' => [
                'min_range' => 1,
                'max_range' => 1000,
            ]]);
            $offset = \filter_var($arguments['offset'] ?? '0', \FILTER_VALIDATE_INT, ['options' => ['min_range' => 0]]);
            if ($rowLimit === false || $offset === false) {
                throw new \InvalidArgumentException(
                    'hub_rows requires limit: 1–1000 and a nonnegative integer offset.',
                );
            }
            if ($inputLimit !== null) {
                $rowLimit = \min($rowLimit, $inputLimit);
            }
            $pagination = http_pagination_offset(
                http_request_option_query('offset'),
                http_request_option_query('limit'),
                \min(100, $rowLimit),
                start_offset: $offset,
                stop_when: http_stop_when_empty_path('rows')->or(http_stop_when_max_results($rowLimit)),
            );
        }

        $frame = data_frame()
            ->read(from_http_paginated($this->client, $request, $pagination))
            ->withEntry('__body', cast(ref('response_body'), type_array()))
            ->withEntry('__row', array_expand(array_get(ref('__body'), $dataset->rowsPath)));

        if ($rowLimit !== null) {
            // The final HTTP page may extend past the requested sample. Cap
            // before user transforms, so aggregations see exactly that sample.
            $frame = $frame->limit($rowLimit);
        }

        foreach (\array_keys($dataset->columns) as $column) {
            // `optional` because the API omits null fields rather than sending
            // them as null, and a bare `array_get` throws on a missing path. A
            // run that succeeded has no `error` key at all; without this, one
            // successful run fails the whole query.
            $frame = $frame->withEntry($column, optional(array_get(ref('__row'), self::source($dataset, $column))));
        }

        // Drop the scaffolding, so a `select()`-free query does not return the
        // whole HTTP exchange alongside the data.
        return $frame->drop(
            ref('__body'),
            ref('__row'),
            ref('response_body'),
            ref('response_headers'),
            ref('response_status_code'),
            ref('response_protocol_version'),
            ref('response_reason_phrase'),
            ref('request_body'),
            ref('request_uri'),
            ref('request_headers'),
            ref('request_protocol_version'),
            ref('request_method'),
        );
    }

    /**
     * Where a column lives inside one record.
     *
     * Runs and spans are already flat. A recorded event nests everything except
     * `event_type` and `data` under `metadata`, and the producing service one
     * level deeper — flattening it here is what lets a query say
     * `ref('service')` instead of `array_get(ref('metadata'), 'source.service')`.
     */
    private static function source(Dataset $dataset, string $column): string
    {
        if ($dataset->name === 'annotation_images') {
            // An image head is `{project, image: {...}, review, ...}`. Reading
            // the record's fields as columns is what lets a query say
            // `ref('group_id')` rather than `array_get(ref('image'), 'group_id')`.
            return match ($column) {
                'review' => $column,
                default => 'image.' . $column,
            };
        }

        if ($dataset->name !== 'events') {
            return $column;
        }

        return match ($column) {
            'event_type', 'data' => $column,
            'service' => 'metadata.source.service',
            'sdk' => 'metadata.source.sdk',
            default => 'metadata.' . $column,
        };
    }
}

File: services/query/contract/aiwatcher_query/catalog.py
"""The catalog every query engine serves, read from the one file that declares it.

`services/query/contract/catalog.json` holds what a dataset *is* — its route, where its
rows sit in a response, which parameter replays its cursor, its columns, its hints and
its `read()` arguments — and every engine loads it: Flow through `CatalogFile.php`, the
Python engines through this module (AW-3). How a dataset is *read* stays in each
engine, because that half is written in each engine's own language.

Read strictly, and in Flow's sentences. A key this module does not know is refused
naming the dataset and the key, because the likeliest one is a near miss, and
`"window": true` quietly leaving `windowed` false is a query that sends its window
nowhere and reads everything. `$comment` is the one key every level skips.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

COMMENT = "$comment"

DATASET_KEYS = (
    "name",
    "grain",
    "description",
    "path",
    "rows_path",
    "cursor_param",
    "requires_run",
    "windowed",
    "columns",
    "hints",
    "parameters",
    "corpus",
)
PARAMETER_KEYS = ("name", "required", "description", "values")


class CatalogError(Exception):
    """The declared catalog cannot be read: a start-up failure, never a request's."""


@dataclass(frozen=True)
class Parameter:
    """A named `read()` argument a dataset's route accepts, beyond `run` and `period`."""

    name: str
    required: bool
    description: str
    values: tuple[str, ...] = ()

    def accepts(self, value: str) -> bool:
        return not self.values or value in self.values


@dataclass(frozen=True)
class Dataset:
    """One named thing a query can read: a route of the aiwatcher API, or a corpus on disk.

    The columns are the contract rather than documentation. Every engine projects exactly
    these, in this order, so what `/query/datasets` shows is what a query can use.
    """

    name: str
    grain: str
    description: str
    columns: dict[str, str]
    path: str = ""
    rows_path: str = ""
    cursor_param: str = ""
    hints: dict[str, str] = field(default_factory=dict)
    #: The route addresses one run, so a read without `run=` has no route to call.
    requires_run: bool = False
    #: The route accepts `window_seconds` and `as_of`. The API refuses a parameter it
    #: does not know, so sending the window to a route without one is a 400.
    windowed: bool = False
    parameters: dict[str, Parameter] = field(default_factory=dict)
    #: A name under `AIWATCHER_CORPUS_DIR` when the rows are part files rather than a
    #: route; `path`, `rows_path` and `cursor_param` then describe nothing.
    corpus: str | None = None

    def explain_unknown_parameter(self, name: str) -> str:
        """What to tell somebody who wrote an argument this route has no idea about."""
        available = (
            "It takes: {}.".format(", ".join(f"{key}=" for key in self.parameters))
            if self.parameters
            else "It takes a dataset name, and period= when it is windowed."
        )
        return f'Dataset "{self.name}" has no read() argument "{name}". {available}'

    def describe(self) -> dict[str, Any]:
        """This dataset as `/query/datasets` lists it — the shape Flow's `datasets()` has."""
        return {
            "name": self.name,
            "aliases": ["default"] if self.name == "runs" else [],
            "grain": self.grain,
            "description": self.description,
            "requires_run": self.requires_run,
            "parameters": [
                {
                    "name": parameter.name,
                    "required": parameter.required,
                    "description": parameter.description,
                    "values": list(parameter.values),
                }
                for parameter in self.parameters.values()
            ],
            "columns": [{"name": name, "type": kind} for name, kind in self.columns.items()],
        }


class Catalog:
    """The datasets this service offers.

    A corpus dataset is offered only where a corpus root is configured: without one it
    names files that are not there, and "no spans" would be an answer about the data
    when the truth is that nobody pointed at any.
    """

    def __init__(self, datasets: dict[str, Dataset], *, corpus_root: Path | None) -> None:
        self._datasets = {
            name: dataset
            for name, dataset in datasets.items()
            if dataset.corpus is None or corpus_root is not None
        }

    @classmethod
    def load(cls, path: Path, *, corpus_root: Path | None) -> Catalog:
        return cls(read_catalog(path), corpus_root=corpus_root)

    def resolve(self, name: str) -> Dataset | None:
        """`default` is `runs`: the grain the explorer shows, and the cheap one."""
        return self._datasets.get("runs" if name == "default" else name)

    def all(self) -> list[Dataset]:
        return list(self._datasets.values())

    def names(self) -> list[str]:
        return list(self._datasets)


def read_catalog(path: Path) -> dict[str, Dataset]:
    """Every dataset the file declares, by name, in the order it declares them."""
    try:
        raw = path.read_text(encoding="utf-8")
    except OSError as error:
        raise CatalogError(
            f"The query catalog is not at {path}. Set AIWATCHER_QUERY_CATALOG to where it is."
        ) from error
    try:
        document = json.loads(raw)
    except json.JSONDecodeError as error:
        raise CatalogError(f"The query catalog at {path} is not JSON: {error}") from error

    version = document.get("version") if isinstance(document, dict) else None
    if (
        not isinstance(document, dict)
        or isinstance(version, bool)
        or version != 1
        or not isinstance(document.get("datasets"), list)
    ):
        raise CatalogError(f"The query catalog at {path} is not version 1 with a list of datasets.")
    _refuse_unknown(document, ("version", "datasets"), f"the catalog at {path}")

    datasets: dict[str, Dataset] = {}
    for entry in document["datasets"]:
        dataset = _dataset(entry, path)
        if dataset.name in datasets:
            raise CatalogError(f'{path} declares "{dataset.name}" twice.')
        datasets[dataset.name] = dataset
    return datasets


def _dataset(entry: object, path: Path) -> Dataset:
    if not isinstance(entry, dict) or not isinstance(entry.get("name"), str):
        raise CatalogError(f"A dataset in {path} has no name.")
    where = f'dataset "{entry["name"]}" in {path}'
    _refuse_unknown(entry, DATASET_KEYS, where)
    return Dataset(
        name=entry["name"],
        grain=_text(entry, "grain", where),
        description=_text(entry, "description", where),
        columns=_table(entry, "columns", where),
        path=_text(entry, "path", where, ""),
        rows_path=_text(entry, "rows_path", where, ""),
        cursor_param=_text(entry, "cursor_param", where, ""),
        hints=_table(entry, "hints", where),
        requires_run=_flag(entry, "requires_run", where),
        windowed=_flag(entry, "windowed", where),
        parameters=_parameters(entry, where),
        corpus=_text(entry, "corpus", where) if "corpus" in entry else None,
    )


def _parameters(entry: dict[str, Any], where: str) -> dict[str, Parameter]:
    declared = entry.get("parameters", [])
    if not isinstance(declared, list):
        raise CatalogError(f'{where}: "parameters" is a list.')
    parameters: dict[str, Parameter] = {}
    for one in declared:
        if not isinstance(one, dict) or not isinstance(one.get("name"), str):
            raise CatalogError(f"{where}: a parameter has no name.")
        at = f'parameter "{one["name"]}" of {where}'
        _refuse_unknown(one, PARAMETER_KEYS, at)
        values = one.get("values", [])
        if not isinstance(values, list) or not all(isinstance(value, str) for value in values):
            raise CatalogError(f'{at}: "values" is a list of strings.')
        parameters[one["name"]] = Parameter(
            name=one["name"],
            required=_flag(one, "required", at),
            description=_text(one, "description", at),
            values=tuple(values),
        )
    return parameters


def _text(entry: dict[str, Any], key: str, where: str, default: str | None = None) -> str:
    value = entry.get(key, default)
    if not isinstance(value, str):
        raise CatalogError(f'{where}: "{key}" is a string, and it is required.')
    return value


def _flag(entry: dict[str, Any], key: str, where: str) -> bool:
    value = entry.get(key, False)
    if not isinstance(value, bool):
        raise CatalogError(f'{where}: "{key}" is true or false.')
    return value


def _table(entry: dict[str, Any], key: str, where: str) -> dict[str, str]:
    """A name-to-text object — columns to their types, or names to hints."""
    value = entry.get(key, {})
    if not isinstance(value, dict) or not all(
        isinstance(text, str) for name, text in value.items() if name != COMMENT
    ):
        raise CatalogError(f'{where}: "{key}" maps names to text.')
    return {name: text for name, text in value.items() if name != COMMENT}


def _refuse_unknown(entry: dict[str, Any], known: tuple[str, ...], where: str) -> None:
    for key in entry:
        if key != COMMENT and key not in known:
            raise CatalogError(
                f'{where} has a key "{key}" this catalog does not know. '
                f"The keys are: {', '.join(known)}."
            )
